### PR TITLE
create staking bucket pool at Greenland height

### PR DESCRIPTION
**Pull request by dustinxie**

### CODE REVIEW

File: action/protocol/staking/bucket_pool.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+}
+
+func (bp *bucketPool) createOrStashBucketPool(ctx context.Context, sm protocol.StateManager) error {
+	blkCtx := protocol.MustGetBlockCtx(ctx)
+	if bp.hu.IsPost(config.Greenland, blkCtx.BlockHeight) {
+		// at Greenland height, create the staking protocol address
+		return accountutil.StoreAccount(sm, bp.addr, bp.pool)
+	}
+	// stash pool total amount to sm
+	return sm.Load(stakingBucketPool, bp.total.Bytes())
```
Comment (dustinxie) [inline action/protocol/staking/bucket_pool.go:158]: similar to candidate center, we store the newest total into workingset's Dock

File: action/protocol/staking/bucket_pool.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+			return nil
+		}
+		return err
+	}
+
+	ser, ok := v.([]byte)
+	if !ok {
+		return errors.Wrap(ErrTypeAssertion, "failed type assertion, expecting []byte")
+	}
+	bp.total.SetBytes(ser)
```
Comment (dustinxie) [inline action/protocol/staking/bucket_pool.go:121]: similar to candidate center, we read the newest total from workingset's Dock

File: action/protocol/staking/bucket_pool.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+		total *big.Int
+	}
+)
+
+func NewBucketPool(addr address.Address) *bucketPool {
+	s := state.EmptyAccount()
+	return &bucketPool{
+		addr:  addr,
+		pool:  &s,
+		total: big.NewInt(0),
```
Comment (yutongp) [inline action/protocol/staking/bucket_pool.go:47]: need implement readstate method: https://github.com/iotexproject/iotex-proto/blob/master/proto/api/read_state.proto#L66
 to read out total amount with corresponding height

File: action/protocol/staking/bucket_pool.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+// in the same way when a bucket are withdrawn, bucket amount is added back to user, but does not come out of any address
+//
+// for better accounting/auditing, we take protocol's address as the 'bucket pool' address
+// 1. at Greenland height we sum up all existing bucket's amount and set the total amount to bucket pool address
+// 2. for future bucket creation/deposit/registration, the amount of staked IOTX token will be added to bucket pool (so
+//    the pool is 'receiving' token)
+// 3. for future bucket withdrawal, the bucket amount will be deducted from bucket pool (so the pool is 'releasing' token)
+
+type (
+	bucketPool struct {
```
Comment (yutongp) [inline action/protocol/staking/bucket_pool.go:33]: shall this be wrapped in candidate state manager?

File: action/protocol/staking/stakingpb/staking.proto [hunk 5]
```diff
@@ -41,3 +41,7 @@ message Candidate {
 message Candidates {
     repeated Candidate candidates = 1;
 }
+
+message TotalAmount {
+    bytes amount = 1;
```
Comment (yutongp) [inline action/protocol/staking/stakingpb/staking.proto:46]: use string for amount

File: action/protocol/staking/bucket_pool.go (hunk 4 above)
Comment (dustinxie) [reply action/protocol/staking/bucket_pool.go:33]: done in latest push

File: action/protocol/staking/candidate_statemanager.go [hunk 6]
```diff
@@ -29,24 +31,34 @@ type (
 		GetByOwner(address.Address) *Candidate
 		GetBySelfStakingIndex(uint64) *Candidate
 		Upsert(*Candidate) error
+		DebitBucketPool(context.Context, *big.Int) error
+		CreditBucketPool(context.Context, *big.Int) error
```
Comment (yutongp) [inline action/protocol/staking/candidate_statemanager.go:35]: should also has a read method for total bucket amount here?

File: action/protocol/staking/candidate_statemanager.go [hunk 7]
```diff
@@ -29,24 +31,34 @@ type (
...
 	candSM struct {
 		protocol.StateManager
 		CandidateCenter
+		*bucketPool
 	}
 )
 
 // NewCandidateStateManager returns a new CandidateStateManager instance
-func NewCandidateStateManager(sm protocol.StateManager, c CandidateCenter) (CandidateStateManager, error) {
+func NewCandidateStateManager(sm protocol.StateManager, c CandidateCenter, pool *bucketPool) (CandidateStateManager, error) {
```
Comment (yutongp) [inline action/protocol/staking/candidate_statemanager.go:47]: I kinda of think that csm should totally own the bucketPool. it is not common a public method has a input with private type. it's better just construct bucketPool within csm creation.

in that case, csm will have a start function. One thing i am not sure is who should call statedb wrapper's start function, is the protocol using it? or factory? @CoderZhi may have some thoughts on this

Comment (dustinxie) [reply action/protocol/staking/candidate_statemanager.go:47]: i know what you mean, in order to do that, we need to change the current way of saving candidate center in the SR's view

currently what's saved in SR's view is the address of the in-memory candidate center object, kinda hacky but works.

we need to serialize the candidate center into bytes, plus the bucket pool's serialized bytes, and store into SR view together

then we can reconstruct the candidate center + bucket pool together by reading SR's view

Comment (dustinxie) [reply action/protocol/staking/candidate_statemanager.go:47]: or easiest way is to move struct TotalAmount into Candidate center struct

File: action/protocol/staking/bucket_pool.go (hunk 3 above)
Comment (dustinxie) [reply action/protocol/staking/bucket_pool.go]: yes, this is a TODO

File: action/protocol/staking/stakingpb/staking.proto (hunk 5 above)
Comment (dustinxie) [reply action/protocol/staking/stakingpb/staking.proto]: done

File: action/protocol/staking/candidate_statemanager.go [hunk 8]
```diff
@@ -16,10 +18,17 @@ import (
 )
 
 type (
-	// CandidateStateManager is candidate manager on top of StateMangaer
+	// CandidateBucketCenter contains candidate center and bucket pool
+	CandidateBucketCenter interface {
+		// TODO: remove CandidateCenter interface, return *candCenter
+		CandCenter() CandidateCenter
+		BucketPool() *bucketPool
```
Comment (yutongp) [inline action/protocol/staking/candidate_statemanager.go:25]: Not normal public interface public method return private struct

File: action/protocol/staking/bucket_pool.go [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+		return errors.Wrap(err, "failed to sync bucket pool")
+	case protocol.ErrNoName:
+		return nil
+	}
+
+	if err := bp.pending.Deserialize(ser); err != nil {
+		return err
+	}
+	fmt.Printf("dirty bytes = %x\n", ser)
+	fmt.Println("===== amount", bp.pending.amount.String())
```
Comment (yutongp) [inline action/protocol/staking/bucket_pool.go:182]: Remove debug code

File: action/protocol/staking/candidate_statemanager.go (hunk 8 above)
Comment (dustinxie) [reply action/protocol/staking/candidate_statemanager.go:25]: change to private interface

File: action/protocol/staking/bucket_pool.go (hunk 3 above)
Comment (dustinxie) [reply action/protocol/staking/bucket_pool.go]: done in latest commit

File: go.mod [hunk 10]
```diff
@@ -17,7 +17,7 @@ require (
 	github.com/iotexproject/go-fsm v1.0.0
 	github.com/iotexproject/go-p2p v0.2.12
 	github.com/iotexproject/go-pkgs v0.1.2-0.20200523040337-5f1d9ddaa8ee
-	github.com/iotexproject/iotex-address v0.2.1
+	github.com/iotexproject/iotex-address v0.2.2-0.20200718221015-84634386838e
```
Comment (yutongp) [inline go.mod:20]: maybe tag a new version for this package?

File: action/protocol/staking/read_state.go [hunk 11]
```diff
@@ -119,6 +121,15 @@ func readStateCandidateByAddress(ctx context.Context, cc CandidateCenter,
 	return c.toIoTeXTypes(), nil
 }
 
+func readStateTotalStakingAmount(ctx context.Context, bp *BucketPool,
+	_ *iotexapi.ReadStakingDataRequest_TotalStakingAmount) (*iotextypes.AccountMeta, error) {
+	meta := iotextypes.AccountMeta{}
+	meta.Address = address.StakingBucketPoolAddr
+	meta.Balance = bp.Total().String()
+	meta.NumActions = bp.Count()
```
Comment (yutongp) [inline action/protocol/staking/read_state.go:129]: maybe don't use numactions this way. we should avoid reuse field which are not the same

Comment (dustinxie) [reply action/protocol/staking/read_state.go:129]: fixed

File: go.mod (hunk 10 above)
Comment (dustinxie) [reply go.mod:20]: done v0.2.2

File: action/protocol/staking/bucket_pool.go [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+}
+
+func (t *totalAmount) Deserialize(data []byte) error {
+	gen := stakingpb.TotalAmount{}
+	if err := proto.Unmarshal(data, &gen); err != nil {
+		return err
+	}
+
+	var ok bool
+	if t.amount, ok = new(big.Int).SetString(gen.Amount, 10); !ok {
```
[the remark this answers is not in the document]
Comment (dustinxie) [reply action/protocol/staking/bucket_pool.go:68]: ok

File: action/protocol/staking/bucket_pool.go [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+
+func (t *totalAmount) AddBalance(amount *big.Int, newBucket bool) {
+	t.amount.Add(t.amount, amount)
+	if newBucket {
+		t.count++
+	}
+}
+
+func (t *totalAmount) SubBalance(amount *big.Int) error {
+	if amount.Cmp(t.amount) == 1 || t.count == 0 {
```
[the remark this answers is not in the document]
Comment (dustinxie) [reply action/protocol/staking/bucket_pool.go:83]: yes, in that case Cmp() == 0. that means it is the last bucket to withdraw, it should be allowed right?

File: action/protocol/staking/bucket_pool.go [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+	}
+
+	// sum up all existing buckets
+	all, err := getAllBuckets(sr)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, v := range all {
+		bp.total.amount.Add(bp.total.amount, v.StakedAmount)
```
[the remark this answers is not in the document]
Comment (dustinxie) [reply action/protocol/staking/bucket_pool.go:117]: v.StakedAmount is created by CreateStake() and store to DB, so i think it's OK to assume it is positive value?
but it's OK, it never hurts to add a check

File: action/protocol/staking/candidate_statemanager.go [hunk 15]
```diff
@@ -29,57 +38,116 @@ type (
...
+}
+
+func (csm *candSM) ContainsSelfStakingBucket(index uint64) bool {
+	return csm.candCenter.ContainsSelfStakingBucket(index)
+}
+
+func (csm *candSM) GetByName(name string) *Candidate {
+	return csm.candCenter.GetByName(name)
+}
+
```
[the remark this answers is not in the document]
Comment (dustinxie) [reply action/protocol/staking/candidate_statemanager.go:131]: the reason we need these is that we discussed early NOT to embed *candCenter, so have to explicitly implement them

File: action/protocol/staking/bucket_pool.go [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+// for better accounting/auditing, we take protocol's address as the 'bucket pool' address
+// 1. at Greenland height we sum up all existing bucket's amount and set the total amount to bucket pool address
+// 2. for future bucket creation/deposit/registration, the amount of staked IOTX token will be added to bucket pool (so
+//    the pool is 'receiving' token)
+// 3. for future bucket withdrawal, the bucket amount will be deducted from bucket pool (so the pool is 'releasing' token)
+
+type (
+	// BucketPool implements the bucket pool
+	BucketPool struct {
+		exist bool
```
[the remark this answers is not in the document]
Comment (dustinxie) [reply action/protocol/staking/bucket_pool.go:43]: replied in chat

File: action/protocol/staking/bucket_pool.go (hunk 9 above)
Comment (dustinxie) [reply action/protocol/staking/bucket_pool.go]: done

**Pull request merged**